README.md
## Rd-WinRM-plugin
This is a [Rundeck Node Execution plugin](http://rundeck.org/docs/plugins-user-guide/node-execution-plugins) that uses WinRM to connect to Windows and execute commands. It uses the [WinRM for Ruby](https://github.com/WinRb/WinRM)Library to provide the WinRM implementation.

Compatible with Rundeck 2.9.x+

## Features
Can run scripts or commands.  
Supports PowerShell, CMD and WQL shells    
Can copy files to Windows  

### Installation

Install Ruby:  
Ubuntu: `apt-get install make ruby ruby-dev`  
CentOS/RHEL: `yum install make ruby ruby-devel`

Install following gems:  
`gem install winrm -v 2.2.3`  
`gem install winrm-fs -v 1.0.2`  
`gem install rubyntlm -v 0.6.2`  

Download from the [releases page](https://github.com/NetDocuments/rd-winrm-plugin/releases).

Copy the `rd-winrm-plugin.zip` to the `libext/` directory for Rundeck. It must be named like `rd-winrm-plugin-x.x.x.zip`. There is no need to restart rundeck.

```bash
RD_WINRM='1.7.0'
curl https://github.com/NetDocuments/rd-winrm-plugin/archive/$RD_WINRM.zip -o /var/lib/rundeck/libext/rd-winrm-plugin-$RD_WINRM.zip
```

Before rundeck can run commands on windows nodes, [configure winrm](https://technet.microsoft.com/en-us/magazine/ff700227.aspx) from an administrative powershell window

    winrm quickconfig

### Configuration
Choose `WinRM Executor` as Default Node Executor  
and `WinRM File Copier` as Default Node File Copier   

Settings:  
`Kerberos Realm`  Put here fqdn of your realm in case your computer is part of AD domain  
`Username` Put here username for negotiate, plaintext or ssl auth  
`Password` Put here password for negotiate, plaintext or ssl auth  
`Auth type` Choose here negotiate, kerberos, plaintext or ssl  
`NOSSL` Choose to disable SSL Validation  
`Allow override` Allow override of variables from job options  
`WinRM transport protocol` Set protocol for winrm transport (Default: http)
`WinRM port` Set port for winrm (Default: 5985/5986 for http/https)  
`Shell` Choose here powershell, cmd or wql  
`WinRM timeout` Put here timeout in seconds (useful for long running commands)  

![](http://cl.ly/1S1D2C070Z1T/Screenshot%202016-01-05%2016.51.53.png)

### Special Behaviour
`Allow Override` parameter gives possibility to set hostname, username and password in job options, not in project. It can be used in case you need to quickly change hostnames (with dropdown list for example) or set username/pass on job level  

- If that parameter set to `hostname` you may use option variable with name `winrmhost` to set hostname
- If that parameter set to `user` you may use `winrmuser` and `winrmpass` to set username/pass
- If that parameter set to `all` you may use all these additional options
- If that parameter set to `none` these options in jobs will be ignored

### Limitations
- Scripts in c:/tmp with .sh extension will be renamed into .ps1, .bat or .wql
- Quotes behaviour can be strange, [Rundeck bug](https://github.com/rundeck/rundeck/issues/602)

### Troubleshooting
You may have some errors like ```WinRM::WinRMAuthorizationError```.  
You can run the following commands on the server to try to solve the problem:

```
winrm set winrm/config/client/auth @{Basic="true"}
winrm set winrm/config/service/auth @{Basic="true"}
winrm set winrm/config/service @{AllowUnencrypted="true"}
```
You can read more about that on issue [#29](https://github.com/WinRb/WinRM/issues/29) on ruby WinRM page

### PR and reporting
PR is highly welcome, we using gitflow for our development process, so please, make them to develop branch.  
If you have some issue please describe steps to reproduce it

### License and Authors
Copyright 2015 NetVoyage Corporation (NetDocuments)

Author: [Volodymyr Babchynskyy](https://github.com/vvchik)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License
is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
## Changelog
### [1.7.0](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.7.0)
- Update winrm and winrm-rf gems
- Rewrite of connection portion of script to support new features of these gems
- Add basic error handling to improve user friendliness of output
- Removed non functioning quote bug fix, Rundeck's documentation is an adequate fix until bug is resolved internally
- Added grouping to UI elements of plugin

### [1.6.2](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.6.2)
- Use Gem winrm 2.x which permit non-administrator session, see
https://github.com/WinRb/WinRM/issues/194
- Use Gem winrm-fs 1.x
- Allow WinRM transport protocol to be specified (HTTP/HTTPS) so
HTTPS can be selected without using 'ssl' authentication type.
- Allow empty defaut user and password at the project level, or
above, so only "overriding" in job options can be use instead

### [1.6.0](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.6.0)
Enables basic auth (GH-34)

### [1.5.1](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.5.1)
Add NTLM/Negotiate authentication type for WinRM File Copier

### [1.5.0](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.5.0)
Add NTLM/Negotiate authentication type [#19](https://github.com/NetDocuments/rd-winrm-plugin/issues/19)
Add check for required winrm/winrm-fs gem versions [#29](https://github.com/NetDocuments/rd-winrm-plugin/issues/29)

### [1.4.1](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.4.1)
Fix HTTPS with self-signed cert [#23](https://github.com/netdocuments/rd-winrm-plugin/issues/23)  
Added support https for ssl connections [#25](https://github.com/NetDocuments/rd-winrm-plugin/pull/25)  

### [1.4.0](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.4.0)
Fix credential leakage [#22](https://github.com/netdocuments/rd-winrm-plugin/issues/22)  
Change shell executions to use non deprecated methods [#24](https://github.com/netdocuments/rd-winrm-plugin/issues/24)  

### [1.3.2](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.3.2)
Fix for error output in case it is not in XML format

### [1.3.1](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.3.1)
Fixes in readme, fix for #3 (WinRM timeout defaults) and #2 (can't modify frozen string)

### [1.3.0](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.3.0)
Extension for copied script depends on context of script execution  

### [1.2.0](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.2.0)
Allow override feature and fix for [rundeck bug](https://github.com/rundeck/rundeck/issues/1421)  

### [1.1.1](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.1.1)
Sanitised passwords in debug mode  

### [1.1.0](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.1.0)
Added possibility to set port for winrm  

### [1.0.0](https://github.com/NetDocuments/rd-winrm-plugin/releases/tag/1.0.0)
Public release, debugging depends on debugging level on job  

### 0.9.0
Fixes in quotes cleanup behaviour, Fixes in quotes cleanup behaviour, trivial rubocop fixes  

### 0.8.0
WinRM timeout managed via config  

### 0.7.0
different shells support (powershell, cmd, wql)  

### 0.6.0
Different authentication types for winrm  

### 0.5.0
convert XML output into readable strings, Password as secure string, cleanup  

### 0.4.0
fix for unix style file manipulation in rundeck on windows, added timeout for log running command  

### 0.3.0
fix for strange rundeck behaviour with quotes  

### 0.2.0
added file copy feature  

### 0.1.0
initial commit  


contents/winrmcp.rb
#!/usr/bin/ruby
gem 'winrm-fs', '= 1.0.2'
require 'winrm-fs'
auth = ENV['RD_CONFIG_AUTHTYPE']
nossl = ENV['RD_CONFIG_NOSSL'] == 'true' ? true : false
if ENV['RD_CONFIG_USER'] # allow empty (default) user (override used)
  user = ENV['RD_CONFIG_USER'].dup # for some reason this string is frozen, so we duplicate it
else
  user =''
end
if ENV['RD_CONFIG_PASS'] # allow empty (default) password (override used)
  pass = ENV['RD_CONFIG_PASS'].dup # for some reason this string is frozen, so we duplicate it
else
  pass = ''
end
host = ENV['RD_NODE_HOSTNAME']
port = ENV['RD_CONFIG_WINRMPORT']
transport = ENV['RD_CONFIG_WINRMTRANSPORT']
shell = ENV['RD_CONFIG_SHELL']
realm = ENV['RD_CONFIG_KRB5_REALM']
override = ENV['RD_CONFIG_ALLOWOVERRIDE']
if ENV['RD_CONFIG_WINRMTIMEOUT']
  timeout = ENV['RD_CONFIG_WINRMTIMEOUT'].to_i
else
  timeout = 60
end
host = ENV['RD_OPTION_WINRMHOST'] if ENV['RD_OPTION_WINRMHOST'] && (override == 'host' || override == 'all')
user = ENV['RD_OPTION_WINRMUSER'].dup if ENV['RD_OPTION_WINRMUSER'] && (override == 'user' || override == 'all')
pass = ENV['RD_OPTION_WINRMPASS'].dup if ENV['RD_OPTION_WINRMPASS'] && (override == 'pass' || override == 'all')

file = ARGV[1]
dest = ARGV[2]
if auth == 'ssl'
  endpoint = "https://#{host}:#{port}/wsman"
else
  endpoint = "#{transport}://#{host}:#{port}/wsman"
end

# Wrapper to fix: "not setting executing flags by rundeck for 2nd file in plugin"
# # https://github.com/rundeck/rundeck/issues/1421
# remove it after issue will be fixed
if File.exist?("#{ENV['RD_PLUGIN_BASE']}/winrmexe.rb") && !File.executable?("#{ENV['RD_PLUGIN_BASE']}/winrmexe.rb")
  File.chmod(0764, "#{ENV['RD_PLUGIN_BASE']}/winrmexe.rb") # https://github.com/rundeck/rundeck/issues/1421
end
#---

# Wrapper for avoid unix style file copying then scripts run
# - not accept chmod call
# - replace rm -f into rm -force
# - auto copying renames file from .sh into .ps1, .bat or .wql in tmp directory
if %r{/tmp/.*\.sh}.match(dest)
  case shell
  when 'powershell'
    dest = dest.gsub(/\.sh/, '.ps1')
  when 'cmd'
    dest = dest.gsub(/\.sh/, '.bat')
  when 'wql'
    dest = dest.gsub(/\.sh/, '.wql')
  end
end
#---

# Build connection options
connections_opts = {
  endpoint: endpoint
}

connections_opts[:operation_timeout] = timeout

case auth
when 'negotiate'
  connections_opts[:transport] = :negotiate
  connections_opts[:user] = user
  connections_opts[:password] = pass
when 'kerberos'
  connections_opts[:transport] = :kerberos
  connections_opts[:realm] = realm
when 'plaintext'
  connections_opts[:transport] = :plaintext
  connections_opts[:user] = user
  connections_opts[:password] = pass
  connections_opts[:disable_sspi] = true
when 'ssl'
  connections_opts[:transport] = :ssl
  connections_opts[:user] = user
  connections_opts[:password] = pass
  connections_opts[:disable_sspi] = true
  connections_opts[:no_ssl_peer_verification] = nossl
else
  fail "Invalid authtype '#{auth}' specified, expected: negotiate, kerberos, plaintext, ssl."
end
#---

# Create session
winrm = WinRM::Connection.new(connections_opts)
file_manager = WinRM::FS::FileManager.new(winrm)
#---

## Upload file to host
begin
  file_manager.upload(file, dest)
rescue HTTPClient::ConnectTimeoutError => e #Capture Timeout on FileCopy (Server Offline)
  if ENV['RD_JOB_LOGLEVEL'] == 'DEBUG'
    STDERR.print "FileCopy failed due to Timeout:\n"
    STDERR.print "Exception Class: #{ e.class.name }\n"
    STDERR.print "Exception Message: #{ e.message }\n"
    STDERR.print "Exception Backtrace: #{ e.backtrace }\n"
    exit 1
  else
    STDERR.print "FileCopy failed due to Timeout: #{ e.class.name }--#{ e.message }\n"
    exit 1
  end
rescue WinRM::WinRMAuthorizationError => e #Capture WinRM Access error (Bad WinRM config)
  if ENV['RD_JOB_LOGLEVEL'] == 'DEBUG'
    STDERR.print "FileCopy failed due to WinRM Access failure:\n"
    STDERR.print "Exception Class: #{ e.class.name }\n"
    STDERR.print "Exception Message: #{ e.message }\n"
    STDERR.print "Exception Backtrace: #{ e.backtrace }\n"
    exit 1
  else
    STDERR.print "FileCopy failed due to WinRM Access failure: #{ e.class.name }--#{ e.message }\n"
    exit 1
  end
rescue => e
  if ENV['RD_JOB_LOGLEVEL'] == 'DEBUG'
    STDERR.print "FileCopy failed due to unhandled exception\n"
    STDERR.print "Exception Class: #{ e.class.name }\n"
    STDERR.print "Exception Message: #{ e.message }\n"
    STDERR.print "Exception Backtrace: #{ e.backtrace }\n"
    exit 1
  else
    STDERR.print "FileCopy failed due to unhandled exception: #{ e.class.name }--#{ e.message }\n"
    exit 1
  end
end
#---


contents/winrmexe.rb
#!/usr/bin/ruby
gem 'winrm', '= 2.2.3'
require 'winrm'
auth = ENV['RD_CONFIG_AUTHTYPE']
nossl = ENV['RD_CONFIG_NOSSL'] == 'true' ? true : false
if ENV['RD_CONFIG_USER'] # allow empty (default) user (override used)
  user = ENV['RD_CONFIG_USER'].dup # for some reason this string is frozen, so we duplicate it
else
  user =''
end
if ENV['RD_CONFIG_PASS'] # allow empty (default) password (override used)
  pass = ENV['RD_CONFIG_PASS'].dup # for some reason this string is frozen, so we duplicate it
else
  pass = ''
end
host = ENV['RD_NODE_HOSTNAME']
port = ENV['RD_CONFIG_WINRMPORT']
transport = ENV['RD_CONFIG_WINRMTRANSPORT']
shell = ENV['RD_CONFIG_SHELL']
realm = ENV['RD_CONFIG_KRB5_REALM']
command = ENV['RD_EXEC_COMMAND']
override = ENV['RD_CONFIG_ALLOWOVERRIDE']
if ENV['RD_CONFIG_WINRMTIMEOUT']
  timeout = ENV['RD_CONFIG_WINRMTIMEOUT'].to_i
else
  timeout = 60
end
host = ENV['RD_OPTION_WINRMHOST'] if ENV['RD_OPTION_WINRMHOST'] && (override == 'host' || override == 'all')
user = ENV['RD_OPTION_WINRMUSER'].dup if ENV['RD_OPTION_WINRMUSER'] && (override == 'user' || override == 'all')
pass = ENV['RD_OPTION_WINRMPASS'].dup if ENV['RD_OPTION_WINRMPASS'] && (override == 'pass' || override == 'all')

if auth == 'ssl'
  endpoint = "https://#{host}:#{port}/wsman"
else
  endpoint = "#{transport}://#{host}:#{port}/wsman"
end

# Wrapper to fix: "not setting executing flags by rundeck for 2nd file in plugin"
# # https://github.com/rundeck/rundeck/issues/1421
# remove it after issue will be fixed
if File.exist?("#{ENV['RD_PLUGIN_BASE']}/winrmcp.rb") && !File.executable?("#{ENV['RD_PLUGIN_BASE']}/winrmcp.rb")
  File.chmod(0764, "#{ENV['RD_PLUGIN_BASE']}/winrmcp.rb")
end
#---

# Wrapper for avoid unix style file copying then scripts run
# - not accept chmod call
# - replace rm -f into rm -force
# - auto copying renames file from .sh into .ps1, .bat or .wql in tmp directory
exit 0 if command.include? 'chmod +x /tmp/'

if command.include? 'rm -f /tmp/'
  shell = 'powershell'
  command = command.gsub(%r{rm -f /tmp/}, 'rm -force /tmp/')
end

if %r{/tmp/.*\.sh}.match(command)
  case shell
  when 'powershell'
    command = command.gsub(/\.sh/, '.ps1')
  when 'cmd'
    command = command.gsub(/\.sh/, '.bat')
  when 'wql'
    command = command.gsub(/\.sh/, '.wql')
  end
end
#---

# Output DEBUG messages
if ENV['RD_JOB_LOGLEVEL'] == 'DEBUG'
  puts 'variables:'
  puts "realm => #{realm}"
  puts "endpoint => #{endpoint}"
  puts "user => #{user}"
  puts 'pass => ********'
  puts "timeout => #{timeout}"
  # puts "pass => #{pass}" # uncomment it for full auth debugging
  puts "command => #{ENV['RD_EXEC_COMMAND']}"
  puts "newcommand => #{command}"
  puts ''

  puts 'ENV:'
  ENV.each do |k, v|
    puts "#{k} => #{v}" if v != pass && k != 'RD_CONFIG_PASS'
    puts "#{k} => ********" if v == pass || k == 'RD_CONFIG_PASS'
    # puts "#{k} => #{v}" if v == pass # uncomment it for full auth debugging
  end
end

def stderr_text(stderr)
  doc = REXML::Document.new(stderr)
  begin
    text = doc.root.get_elements('//S').map(&:text).join
    text.gsub(/_x(\h\h\h\h)_/) do
      code = Regexp.last_match[1]
      code.hex.chr
    end
  rescue
    return stderr
  end
end

# Build connection options
connections_opts = {
  endpoint: endpoint
}

connections_opts[:operation_timeout] = timeout

case auth
when 'negotiate'
  connections_opts[:transport] = :negotiate
  connections_opts[:user] = user
  connections_opts[:password] = pass
when 'kerberos'
  connections_opts[:transport] = :kerberos
  connections_opts[:realm] = realm
when 'plaintext'
  connections_opts[:transport] = :plaintext
  connections_opts[:user] = user
  connections_opts[:password] = pass
  connections_opts[:disable_sspi] = true
when 'ssl'
  connections_opts[:transport] = :ssl
  connections_opts[:user] = user
  connections_opts[:password] = pass
  connections_opts[:disable_sspi] = true
  connections_opts[:no_ssl_peer_verification] = nossl
else
  fail "Invalid authtype '#{auth}' specified, expected: negotiate, kerberos, plaintext, ssl."
end
#---

# Create and connect session
winrm = WinRM::Connection.new(connections_opts)
shell_session = nil
case shell
when 'powershell'
  begin
    shell_session = winrm.shell(:powershell)
    result = shell_session.run(command)
  rescue => e
    shell_session = nil
    if ENV['RD_JOB_LOGLEVEL'] == 'DEBUG'
      STDERR.print "ScriptExecution failed due unhandled exception:\n"
      STDERR.print "Exception Class: #{ e.class.name }\n"
      STDERR.print "Exception Message: #{ e.message }\n"
      STDERR.print "Exception Backtrace: #{ e.backtrace }\n"
      exit 1
    else
      STDERR.print "ScriptExecution failed due unhandled exception: #{ e.class.name }--#{ e.message }\n"
      exit 1
    end
  end
when 'cmd'
  begin
    shell_session = winrm.shell(:cmd)
    result = shell_session.run(command)
  rescue => e
    shell_session = nil
    if ENV['RD_JOB_LOGLEVEL'] == 'DEBUG'
      STDERR.print "ScriptExecution failed due unhandled exception:\n"
      STDERR.print "Exception Class: #{ e.class.name }\n"
      STDERR.print "Exception Message: #{ e.message }\n"
      STDERR.print "Exception Backtrace: #{ e.backtrace }\n"
      exit 1
    else
      STDERR.print "ScriptExecution failed due unhandled exception: #{ e.class.name }--#{ e.message }\n"
      exit 1
    end
  end
when 'wql'
  result = winrm.run_wql(command)
end
#---

# Organise output for return to Rundeck
if shell_session != nil
  STDERR.print stderr_text(result.stderr) if result.stderr != ''
  STDOUT.print result.stdout
  exit result.exitcode if result.exitcode !=0
else # WQL
  STDOUT.print result
end
#---
